core/include/detray/intersection/intersection_kernel.hpp

```diff
@@ -7,105 +7,123 @@
- *
- * @return an intersection struct (invalid if no intersection was found)
+/** A functor to add all valid intersections between the trajectory and surface
  */
-template <typename track_t, typename transform_t, typename mask_container_t,
-          typename mask_range_t, unsigned int first_mask_id,
-          unsigned int... remaining_mask_ids>
-DETRAY_HOST_DEVICE inline auto unroll_intersect(
-    const track_t &track, const transform_t &ctf, const mask_container_t &masks,
-    const mask_range_t &mask_range,
-    const typename mask_container_t::id_type mask_id, dindex volume_index,
-    std::integer_sequence<unsigned int, first_mask_id, remaining_mask_ids...>
-    /*available_ids*/) {
-
-    // Pick the first one for interseciton
-    if (mask_id == first_mask_id) {
-
-        auto &mask_group =
-            masks.template group<mask_container_t::to_id(first_mask_id)>();
-
-        // Check all masks of this surface for intersection
+struct intersection_initialize {
+
+    using output_type = std::size_t;
+
+    /** Operator function to initalize intersections
+     *
+     * @tparam mask_group_t is the input mask group type found by variadic
+     * unrolling
+     * @tparam is_container_t is the intersection container type
+     * @tparam traj_t is the input trajectory type (e.g. ray or helix)
+     * @tparam surface_t is the input surface type
+     * @tparam transform_container_t is the input transform store type
+     *
+     * @param mask_group is the input mask group
+     * @param is_container is the intersection container to be filled
+     * @param traj is the input trajectory
+     * @param surface is the input surface
+     * @param contextual_transforms is the input transform container
+     * @param edge_tolerance is the tolerance for mask size
+     *
+     * @return the number of valid intersections
+     */
+    template <typename mask_group_t, typename is_container_t, typename traj_t,
+              typename surface_t, typename transform_container_t>
+    DETRAY_HOST_DEVICE inline output_type operator()(
+        const mask_group_t &mask_group, is_container_t &is_container,
+        const traj_t &traj, const surface_t &surface,
+        const transform_container_t &contextual_transforms,
+        const scalar edge_tolerance = 0.) const {
+
+        int count = 0;
+
+        const auto &mask_range = surface.mask_range();
+        const auto &ctf = contextual_transforms[surface.transform()];
+
+        // Run over the masks belonged to the surface
         for (const auto &mask : range(mask_group, mask_range)) {
-            auto sfi =
-                std::move(mask.intersector().intersect(ctf, track, mask));
 
-            if (sfi.status == intersection::status::e_inside) {
-                sfi.index = volume_index;
-                return sfi;
+            auto sfi = std::move(mask.intersector()(
+                traj, mask, ctf, edge_tolerance, traj.overstep_tolerance()));
```

Comment:
Just as a comment: 'edge' in the rest of detray means an edge in a graph (see volume graph) and not so much a geometric edge of a surface. But since the graph is only used in the tests, it should not be too confusing to leave it like this in the core. Some of that graph terminology might have seeped into the masks, I agree, that should probably be removed at some point

Reply:
I think that confusion can be tolerated...

Reply:
Ok, but we should still rename the e.g. typedefs in the detector metadata in this case?

Reply:
We can name it `link_type`. But I don't know if `link_type` needs to be templated. Maybe you can explain when the link type size can be different.

Reply:
```link_type``` sounds good. It is true that a single mask must not lead into multiple volumes, but if we want to check multiple masks at the same time, every one needs a volume link (what is currently sometimes called an edge, while the link to the surfaces finder will probably be moved to the mother volume instead). Then the volume link would become a vector itself. It could, however, be that in a vectorized version, I need specialized masks altogether, then we can remove the template from the mask as far as I can see

Reply:
Alright. I just worried that potential users would get confused by too many template parameters in surface, volume, mask. It was actually a big huddle to me when I started working on this project.  It's not a priority but we should think of how to make the user interface easier.

Reply:
That's true, the learning curve for new developers is steep and the doc comments are also not too helpful currently...